File: testdata/fixtures/src/sample.r
# sample.r — comprehensive R syntax fixture for parser testing.
# Covers: vectors, lists, data frames, functions, closures, S3/S4/R5 classes,
# environments, apply family, pipe operators, error handling, regex,
# tidyverse-style patterns, generics, recursion, metaprogramming.

# --------------------------------------------------------------------------- #
# Vectors & basic operations
# --------------------------------------------------------------------------- #

int_vec    <- 1:20
dbl_vec    <- c(1.5, 2.7, 3.14, NA, Inf)
chr_vec    <- c("alpha", "beta", "gamma")
lgl_vec    <- c(TRUE, FALSE, TRUE, NA)
cplx_vec   <- c(1+2i, 3-4i)

MAX_RETRIES <- 3L
PI_APPROX   <- 22 / 7

# Named vector
scores <- c(alice = 95, bob = 87, carol = 92)

# Sequence helpers
evens   <- seq(0, 100, by = 2)
five_pts <- seq(0, 1, length.out = 5)

# Vectorised operations (no loops needed)
normalized <- (int_vec - mean(int_vec)) / sd(int_vec)
above_mean <- int_vec[int_vec > mean(int_vec)]

# --------------------------------------------------------------------------- #
# Lists & nested structures
# --------------------------------------------------------------------------- #

person <- list(
  name    = "Alice",
  age     = 30L,
  hobbies = c("reading", "cycling"),
  address = list(city = "Berlin", country = "DE")
)

person$address$city        # $ accessor
person[["name"]]           # [[ accessor
person["age"]              # single-bracket returns list

# --------------------------------------------------------------------------- #
# Functions & closures
# --------------------------------------------------------------------------- #

make_adder <- function(n) {
  force(n)           # capture n now (avoid lazy-eval pitfall)
  function(x) x + n
}

add5 <- make_adder(5)

safe_log <- function(x, base = exp(1)) {
  if (any(x <= 0, na.rm = TRUE)) stop(paste0("x must be positive, got: ", x))
  log(x, base)
}

retry <- function(fn, times = MAX_RETRIES, ...) {
  last_err <- NULL
  for (i in seq_len(times)) {
    tryCatch(
      return(fn(...)),
      error = function(e) { last_err <<- e }
    )
  }
  stop(paste0("failed after ", times, " attempts: ", conditionMessage(last_err)))
}

# Recursive function
fibonacci <- function(n) {
  if (n <= 1L) return(n)
  fibonacci(n - 1L) + fibonacci(n - 2L)
}

fibonacci_vec <- Vectorize(fibonacci)

# --------------------------------------------------------------------------- #
# Apply family
# --------------------------------------------------------------------------- #

squares      <- sapply(1:10, function(x) x^2)
cube_list    <- lapply(1:5, function(x) x^3)
row_means    <- apply(matrix(1:12, nrow = 3), 1, mean)
named_double <- vapply(chr_vec, nchar, integer(1))

# --------------------------------------------------------------------------- #
# Data frames
# --------------------------------------------------------------------------- #

df <- data.frame(
  id     = 1:5,
  name   = c("Alice", "Bob", "Carol", "Dave", "Eve"),
  score  = c(92, 85, 78, 95, 88),
  passed = c(TRUE, TRUE, FALSE, TRUE, TRUE),
  stringsAsFactors = FALSE
)

# Subsetting
top_scorers   <- df[df$score >= 90, ]
name_and_score <- df[, c("name", "score")]

# Adding columns
df$grade <- ifelse(df$score >= 90, "A",
              ifelse(df$score >= 80, "B", "C"))

# Aggregate
avg_by_grade <- aggregate(score ~ grade, data = df, FUN = mean)

# --------------------------------------------------------------------------- #
# Pipe operator (|> native, R 4.1+)
# --------------------------------------------------------------------------- #

result <- c(3, 1, 4, 1, 5, 9, 2, 6) |>
  sort() |>
  unique() |>
  rev()

# --------------------------------------------------------------------------- #
# String manipulation
# --------------------------------------------------------------------------- #

greeting  <- sprintf("Hello, %s! You scored %d.", "Alice", 95)
upper     <- toupper(greeting)
words     <- strsplit("one two three", " ")[[1]]
matched   <- regmatches("2026-06-04", regexpr("\\d{4}-\\d{2}-\\d{2}", "2026-06-04"))
replaced  <- gsub("([aeiou])", "[\\1]", "hello world", perl = TRUE)

# --------------------------------------------------------------------------- #
# S3 Classes
# --------------------------------------------------------------------------- #

new_point <- function(x, y) {
  structure(list(x = x, y = y), class = "Point")
}

print.Point <- function(p, ...) {
  cat(sprintf("Point(%.2f, %.2f)\n", p$x, p$y))
}

distance.Point <- function(p, other) {
  sqrt((p$x - other$x)^2 + (p$y - other$y)^2)
}

`+.Point` <- function(a, b) new_point(a$x + b$x, a$y + b$y)

is_origin <- function(p) UseMethod("is_origin")
is_origin.Point <- function(p) p$x == 0 && p$y == 0

p1 <- new_point(3, 4)
p2 <- new_point(0, 0)

# --------------------------------------------------------------------------- #
# S4 Classes
# --------------------------------------------------------------------------- #

setClass("Animal", representation(
  name      = "character",
  weight_kg = "numeric",
  status    = "character"
))

setGeneric("speak", function(animal, ...) standardGeneric("speak"))
setGeneric("weight_lbs", function(animal) standardGeneric("weight_lbs"))

setMethod("speak", "Animal", function(animal, ...) {
  cat(animal@name, "says ...\n")
})

setMethod("weight_lbs", "Animal", function(animal) {
  animal@weight_kg * 2.205
})

setClass("Dog", contains = "Animal", representation(breed = "character"))

setMethod("speak", "Dog", function(animal, ...) {
  cat(animal@name, "says woof!\n")
})

rex <- new("Dog", name = "Rex", weight_kg = 30, status = "active", breed = "Lab")

# --------------------------------------------------------------------------- #
# R5 / Reference Classes
# --------------------------------------------------------------------------- #

Counter <- setRefClass("Counter",
  fields = list(count = "numeric"),
  methods = list(
    initialize = function(start = 0) {
      count <<- start
    },
    increment = function(by = 1) {
      count <<- count + by
    },
    reset = function() {
      count <<- 0
    },
    show = function() {
      cat("Counter:", count, "\n")
    }
  )
)

ctr <- Counter$new(10)

# --------------------------------------------------------------------------- #
# Environments & metaprogramming
# --------------------------------------------------------------------------- #

e <- new.env(parent = emptyenv())
assign("x", 42, envir = e)
get("x", envir = e)

# Quote & eval
expr <- quote(1 + 2 * 3)
eval(expr)

substitute_demo <- function(x) {
  deparse(substitute(x))
}

# --------------------------------------------------------------------------- #
# Error handling
# --------------------------------------------------------------------------- #

safe_divide <- function(a, b) {
  tryCatch(
    {
      if (b == 0) stop(paste0("division by zero: ", a, " / ", b))
      a / b
    },
    error   = function(e) { warning(conditionMessage(e)); NA_real_ },
    warning = function(w) { message("warn: ", conditionMessage(w)); NA_real_ },
    finally = { /* always runs */ }
  )
}

withCallingHandlers(
  log(-1),
  warning = function(w) {
    message("Caught warning: ", conditionMessage(w))
    invokeRestart("muffleWarning")
  }
)

# --------------------------------------------------------------------------- #
# Entry point (when run as script)
# --------------------------------------------------------------------------- #

if (!interactive()) {
  cat("Fibonacci(10) =", fibonacci(10), "\n")
  print(p1)
  cat("Distance:", distance.Point(p1, p2), "\n")
  speak(rex)
  ctr$increment(5)
  ctr$show()
}

File: testdata/fixtures/src/sample.lua
--[[
  sample.lua — comprehensive Lua syntax fixture for parser testing.
  Covers: tables, metatables, OOP, closures, coroutines, iterators,
  error handling, modules, varargs, string patterns, bit ops, environments.
]]

-- -------------------------------------------------------------------------- --
-- Constants
-- -------------------------------------------------------------------------- --

local MAX_RETRIES = 3
local DEFAULT_TIMEOUT = 30.0
local GREETING = "hello"

-- -------------------------------------------------------------------------- --
-- Basic types & operators
-- -------------------------------------------------------------------------- --

local int_val  = 42
local float_val = 3.14
local big_int  = 0x7FFFFFFF
local str_val  = "hello, world"
local long_str = [[
  multi-line
  string literal
]]
local bool_t   = true
local null_val = nil

-- Arithmetic & logical
local result = (int_val + float_val) * 2 - 1
local concat = str_val .. " " .. tostring(int_val)
local len    = #str_val
local band   = int_val & 0xFF     -- bitwise AND (Lua 5.3+)
local bor    = int_val | 0x100
local bxor   = int_val ~ 0x0F
local bnot   = ~int_val
local lsh    = int_val << 2
local rsh    = int_val >> 1

-- -------------------------------------------------------------------------- --
-- Tables (arrays and dicts)
-- -------------------------------------------------------------------------- --

local arr = {10, 20, 30, 40, 50}
local dict = { name = "Alice", age = 30, active = true }
local mixed = { 1, 2, key = "value", nested = { a = 1, b = 2 } }

arr[#arr + 1] = 60              -- append
table.insert(arr, 70)
table.remove(arr, 1)
table.sort(arr, function(a, b) return a > b end)

-- -------------------------------------------------------------------------- --
-- Functions
-- -------------------------------------------------------------------------- --

local function add(a, b)
  return a + b
end

local multiply = function(x, y) return x * y end

-- Multiple return values
local function min_max(t)
  local lo, hi = t[1], t[1]
  for _, v in ipairs(t) do
    if v < lo then lo = v end
    if v > hi then hi = v end
  end
  return lo, hi
end

-- Varargs
local function sum(...)
  local total = 0
  for _, v in ipairs({...}) do total = total + v end
  return total
end

-- Default arguments via "or"
local function greet(name, greeting)
  greeting = greeting or GREETING
  return greeting .. ", " .. name .. "!"
end

-- Closures
local function make_counter(start)
  local count = start or 0
  return {
    increment = function() count = count + 1 end,
    decrement = function() count = count - 1 end,
    value     = function() return count end,
  }
end

-- -------------------------------------------------------------------------- --
-- Metatables & OOP
-- -------------------------------------------------------------------------- --

local Animal = {}
Animal.__index = Animal

function Animal.new(name, weight)
  local self = setmetatable({}, Animal)
  self.name   = name
  self.weight = weight
  return self
end

function Animal:speak()
  return self.name .. " says ..."
end

function Animal:__tostring()
  return ("Animal(%s, %.1fkg)"):format(self.name, self.weight)
end

function Animal:__lt(other)
  return self.weight < other.weight
end

-- Inheritance
local Dog = setmetatable({}, { __index = Animal })
Dog.__index = Dog

function Dog.new(name, weight, breed)
  local self = Animal.new(name, weight)
  setmetatable(self, Dog)
  self.breed  = breed
  self.tricks = {}
  return self
end

function Dog:speak()
  return self.name .. " says woof!"
end

function Dog:learn(trick)
  table.insert(self.tricks, trick)
  return self    -- fluent interface
end

function Dog:perform()
  local out = {}
  for _, t in ipairs(self.tricks) do
    out[#out + 1] = self.name .. " performs " .. t
  end
  return out
end

-- -------------------------------------------------------------------------- --
-- Iterators & coroutines
-- -------------------------------------------------------------------------- --

local function fibonacci()
  local a, b = 0, 1
  return function()
    local v = a
    a, b = b, a + b
    return v
  end
end

local function range(from, to, step)
  step = step or 1
  return function(_, i)
    i = i + step
    if i <= to then return i end
  end, nil, from - step
end

-- Coroutine-based generator
local function fibonacci_co()
  return coroutine.wrap(function()
    local a, b = 0, 1
    while true do
      coroutine.yield(a)
      a, b = b, a + b
    end
  end)
end

-- -------------------------------------------------------------------------- --
-- Error handling
-- -------------------------------------------------------------------------- --

local function safe_call(fn, ...)
  local ok, result = pcall(fn, ...)
  if not ok then
    return nil, result   -- result holds error message
  end
  return result
end

local function safe_divide(a, b)
  if b == 0 then
    error(("division by zero: %d / %d"):format(a, b), 2)
  end
  return a / b
end

local val, err = safe_call(safe_divide, 10, 0)
if err then io.stderr:write("error: " .. err .. "\n") end

-- xpcall with traceback
local function with_traceback(fn, ...)
  return xpcall(fn, function(e)
    return e .. "\n" .. debug.traceback("", 2)
  end, ...)
end

-- -------------------------------------------------------------------------- --
-- String patterns
-- -------------------------------------------------------------------------- --

local date = "2026-06-04"
local year, month, day = date:match("(%d%d%d%d)-(%d%d)-(%d%d)")

local words = {}
for w in ("one two three four"):gmatch("%S+") do
  words[#words + 1] = w
end

local snake = ("CamelCaseName"):gsub("%u", function(c)
  return "_" .. c:lower()
end):gsub("^_", "")

-- -------------------------------------------------------------------------- --
-- Modules (returns a table)
-- -------------------------------------------------------------------------- --

local M = {}

M.VERSION = "1.0.0"

function M.greet(name) return greet(name) end
function M.fibonacci() return fibonacci() end

M.Animal = Animal
M.Dog    = Dog

-- -------------------------------------------------------------------------- --
-- Entry point
-- -------------------------------------------------------------------------- --

local dog = Dog.new("Rex", 30, "Labrador")
dog:learn("sit"):learn("shake")
print(dog:speak())
for _, line in ipairs(dog:perform()) do print(line) end

local fib = fibonacci_co()
local fibs = {}
for _ = 1, 10 do fibs[#fibs + 1] = fib() end
print(table.concat(fibs, ", "))

local lo, hi = min_max({3, 1, 4, 1, 5, 9, 2, 6})
print(("min=%d  max=%d"):format(lo, hi))

return M
